SD-256 enable colored output by default on unix

gourlaysama opened the pull request:
This enables `-Dscala.color` by default when running via the shell
script and when stdout is a terminal.

Can be overriden on the CLI or via `JAVA_OPTS`:

```sh
$ qbin/scala -nc -e 'println(util.Properties.propOrFalse("scala.color"))'
true
$ qbin/scala -nc -e 'println(util.Properties.propOrFalse("scala.color"))' | cat
false
$ qbin/scala -nc -Dscala.color=false -e 'println(util.Properties.propOrFalse("scala.color"))'
false
$ JAVA_OPTS="-Dscala.color=false" qbin/scala -nc -e 'println(util.Properties.propOrFalse("scala.color"))'
false
```

adriaanm commented:
Sweet! Thanks! Anyone out there familiar with unix gotchas like this? I'm not qualified to review, but would love to get this in!

dwijnand commented on src/compiler/scala/tools/ant/templates/tool-unix.tmpl:
```diff
@@ -127,6 +127,11 @@ fi
+# enable colored output by default
+# we prepend it to JAVA_OPTS so users can override it on the
+# command line or via JAVA_OPTS
+[[ -t 1 ]] && JAVA_OPTS="-Dscala.color=true $JAVA_OPTS"
```
Rather than mess directly with JAVA_OPTS I would follow the precendence set by `-Dscala.usejavacp`/`OVERRIDE_USEJAVACP`

gourlaysama replied:
I originally didn't want to clutter the command line to run with one more variable, but you're right, it's cleaner that way. Done.

smarter commented:
That only solves part of the problem, it would also be nice to get colors when running `sbt console` for example, for this the default value of `scala.color` should be `auto` which would behave like `grep --color=auto` and many other tools, but this means that the terminal detection must be done on the JVM.

smarter commented:
`System.console` seems to be the way to do that: https://stackoverflow.com/questions/1403772/how-can-i-check-if-a-java-programs-input-output-streams-are-connected-to-a-term

gourlaysama commented:
The original reason I did it in the bash script was that we only want to activate it by default on unix-like things.

Hmm, having an `auto` mode that tests `System.console` would indeed make it work all on its own in sbt *and* with the bash runner. Then I can add a `-Dscala.color=false` default in the bat runner for windows.
Maybe that's a better solution?

smarter commented:
I think the Windows console supports color now, though I have no idea how to test for that: https://blogs.msdn.microsoft.com/commandline/2016/09/22/24-bit-color-in-the-windows-console/ Simply disabling color in the batch script might be enough.

smarter commented:
Or always disable it the platform is windows: http://www.scala-lang.org/api/2.12.0/scala-compiler/scala/tools/nsc/Properties$.html#isWin:Boolean

gourlaysama commented:
> Or always disable if the platform is windows

Oh, right, I'll just do that.

> I think the Windows console supports color now, though I have no idea how to test for that

After [a bit of digging][2]: it is enabled by default when used from the Windows Subsystem for Linux, but not for normal windows console programs run directly or from `cmd` or powershell. `cmd.exe` does enable it for itself, so it works within a `.bat` script, but `java.exe` doesn't, and that property isn't inherited by child processes for compatibility reasons. So no color out of the box for the normal usecase. We could enable it, but it is [a native API][1], so JNI and all that... let's just keep it disabled on Windows.

[1]: https://msdn.microsoft.com/en-us/library/windows/desktop/ms686033%28v=vs.85%29.aspx
[2]: https://github.com/Microsoft/BashOnWindows/issues/1173#issuecomment-254250445

gourlaysama committed:
SD-256 enable colored output by default on unix

`scala.color` now has 3 states: `true`, `false` and `auto`
(the default). `auto` allows colors if not on windows and if the shell
is interactive (as in, both stdin and stdout are a tty).

The autodetect works as expected when run via SBT too, and it can always
be overriden on the CLI or via JAVA_OPTS.

gourlaysama commented:
@smarter done. Now it works in SBT too :)

I moved the property somewhere more central since it is already used in two places. The next step would be to use it for a few more things, like the default `ConsoleReporter`, and maybe more things in the repl.

gourlaysama commented:
```
library/mima failed: java.lang.RuntimeException: MiMa failed with exit code 2
```
😦

adriaanm commented:
Awesome! Yeah, scala's package-private can't be represented in byte code. Will need whitelisting :-/

lrytz commented:
LGTM - needs a rebase unfortunately

gourlaysama commented:
Rebased.

lrytz commented:
/rebuild - interesting failure

```
Caused by: sbt.ForkMain$ForkError: java.lang.ArrayIndexOutOfBoundsException: 15
	at java.util.ArrayList.add(ArrayList.java:459)
	at sbt.ForkMain$Run$2$1.handle(ForkMain.java:294)
	at scala.tools.partest.sbt.SBTRunner$$anon$1.onFinishTest(SBTRunner.scala:70)
	at scala.tools.partest.nest.SuiteRunner.runTest(Runner.scala:781)
	at scala.tools.partest.nest.SuiteRunner.$anonfun$runTestsForFiles$2(Runner.scala:788)
	at scala.tools.partest.package$$anon$2.call(package.scala:135)
```

adriaanm commented:
🎉

SethTisue commented:
> Now it works in SBT too

confirmed. 🙌

som-snytt commented:
Just noticed my scripts with `-Dscala.color` stopped working. That is, it's no longer taken as `-Dscala.color=true` or even `auto`.
